File: src/test/java/uk/laxd/dndSimulator/action/TurnTest.java
package uk.laxd.dndSimulator.action;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import uk.laxd.dndSimulator.character.Character;
import uk.laxd.dndSimulator.character.CharacterClass;
import uk.laxd.dndSimulator.event.*;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

import static org.junit.Assert.*;

@RunWith(MockitoJUnitRunner.class)
public class TurnTest {

    private Character character;
    private Character target;

    private TargetSelector targetSelector;

    private EncounterEventFactory eventFactory = new EncounterEventFactory();

    private EventLogger eventLogger = new SimpleEventLogger();

    @Mock
    private ActionResolver actionResolver;

    @Before
    public void setUp() throws Exception {
        this.character = new Character("Steve", CharacterClass.BARBARIAN, 20);
        this.target = new Character("Alan", CharacterClass.GENERIC_CHARACTER, 20);

        this.targetSelector = new SimpleTargetSelector(target);
    }

    @Test
    public void testNoTargetResultsInNoHit() throws Exception {
        this.targetSelector = new SimpleTargetSelector(null);
        Turn turn = new Turn(eventFactory, eventLogger, actionResolver, character, targetSelector);

        turn.doTurn();

        assertTrue(eventLogger.getEvents().stream()
                .noneMatch(e -> e.getType() == EncounterEventType.MELEE_ATTACK));
    }

    @Test
    public void testTurnStartEventIsLoggedAtStartOfTurn() throws Exception {
        Turn turn = new Turn(eventFactory, eventLogger, actionResolver, character, targetSelector);

        turn.doTurn();

        assertTrue(eventLogger.getEvents().size() > 0);
        assertEquals(EncounterEventType.TURN_START, eventLogger.getEvents().iterator().next().getType());
    }

}

File: app/src/test/kotlin/action/TurnTest.kt
package uk.laxd.dndSimulator.action

import io.mockk.every
import io.mockk.justRun
import io.mockk.mockk
import org.junit.Assert.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import uk.laxd.dndSimulator.character.Character
import uk.laxd.dndSimulator.event.*

class TurnTest {
    private var character: Character? = null
    private var target: Character? = null
    private var targetSelector: TargetSelector? = null
    private val eventLogger: EventLogger = EventLogger.instance
    private val actionFactory = ActionFactory()
    private val actionResolver: ActionResolver = mockk()

    @BeforeEach
    fun setUp() {
        character = Character("Steve", "Team A")
        target = Character("Alan", "Team B")
        targetSelector = SimpleTargetSelector(listOf(character!!, target!!))

        every { actionResolver.resolve(any()) }.returns(Unit)
    }

    @Test
    fun testNoTargetResultsInNoHit() {
        targetSelector = SimpleTargetSelector(listOf())
        val turn = Turn(actionFactory, actionResolver, character!!, targetSelector!!)
        turn.doTurn()
        assertTrue(eventLogger.events.stream()
            .noneMatch { e -> e.type === EncounterEventType.MELEE_ATTACK })
    }

    @Test
    fun testTurnStartEventIsLoggedAtStartOfTurn() {
        val turn = Turn(actionFactory, actionResolver, character!!, targetSelector!!)

        justRun { actionResolver.resolve(any()) }

        turn.doTurn()
        assertTrue(eventLogger.events.isNotEmpty())
        assertEquals(EncounterEventType.TURN_START, eventLogger.events.iterator().next().type)
    }
}